### PR TITLE
Create Unique TabIds and _always_ require them for actions [WIP]

### LINKED ISSUE #593: Create Unique TabIds and _always_ require them for actions
## Current Behavior

We still assume an intended tab for some actions via `activeTab` (some efforts were made to improve this in #567).

Ideally we should always be passing a tab identifier to actions, so that it's _definitively_ occurring with the tab we want.

## Possible Solution

- [ ] Generate a unique tabId for each tab on creation.
- [ ] _always_ use this for identifying tabs in actions.
- [ ] Remove _getActiveTab_ from actions that are not setting a _new_ activeTab.

### PR DESCRIPTION
**Pull request by manavbp**
Resolves #593

### CODE REVIEW

File: app/actions/winodw_actions.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export const {
+    WindowaddTab
+    , WindowsetActiveTab
+    , WindowcloseTab
+    , WindowreopenTab
+} = createActions(
+    TYPES.WINDOW_ADD_TAB
+    , TYPES.WINDOW_SET_ACTIVE_TAB
+    , TYPES.WINDOW_CLOSE_TAB
+    , TYPES.WINDOW_REOPEN_TAB
```
Comment (joshuef) [inline app/actions/winodw_actions.ts]: window / open/close menu?

File: app/reducers/tabs.ts [hunk 2]
```diff
@@ -1,263 +1,94 @@
-
-import { remote, shell, webContents } from 'electron';
+/* eslint-disable */
```
Comment (joshuef) [inline app/reducers/tabs.ts]: this disabled it for the entire file, I think :(

File: app/reducers/windows.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+/**
+ * Get the current window's webcontents Id. Defaults to `1` if none found.
+ * @return { Integer } WebContents Id of the curremt BrowserWindow webcontents.
+ */
+const getCurrentWindowId = () =>
+{
+    let currentWindowId = 1; // for testing
+    if ( remote )
+    {
+        currentWindowId = remote.getCurrentWindow().webContents.id;
```
Comment (joshuef) [inline app/reducers/windows.ts]: consider how this will work with tests (where there is not necessarily a window)

File: app/reducers/windows.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        currentWindowId = currentWindow.id;
+    }
+    return currentWindowId;
+};
+
+
+const windowAddTab = (state, tab) => 
+{
+    const openWindows = { ...state.openWindows };
+    const targetwindowId = tab.windowId || getCurrentWindowId();
```
Comment (joshuef) [inline app/reducers/windows.ts]: Maybe we should not have this fallback of `getCurrentWindowId`, and _enforce_ passing of a window ID? That might prevent bugs

File: app/reducers/windows.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    if (openWindows[targetwindowId]) {
+      openWindows[targetwindowId].tabs.push(tabId);
+      newWindow = {
+        ...openWindows[targetwindowId],
+        activeTab: tabId,
+        tabs: openWindows[targetwindowId].tabs
+      };
+    } 
+    else if(openWindows[targetwindowId] === undefined){
+      newWindow = {
```
Comment (joshuef) [inline app/reducers/windows.ts]: I'm not sure if we should be adding open windows here, or throwing an error?

**Comment by manavbp** [comment]
> I guess this is just one of many commits for this PR too, right @manavbp ? The actually updating of code to do this will follow?
Yep, So that the PR can be broken down into a few smaller commits!

File: __tests__/reducers/windows.spec.ts [hunk 6]
```diff
@@ -44,16 +45,74 @@ describe( 'windows reducer', () =>
...
+                closedWindows : 
+                {
+                    '1': 
+                    {
+                        closedtabs : [],
+                        lastActiveTabs : []
+                    }
+                }
+            }
+            const addTabNext = windows(JSON.parse(JSON.stringify(state)),{
```
Comment (joshuef) [inline __tests__/reducers/windows.spec.ts]: You're creating a new state just above. I don't think the parse/stringify combo is needed here...?

File: app/reducers/windows.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+  return newState;
+};
+
+const closetab = (state,tab) =>
+{
+  const targetWindow = tab.windowId;
+  const openWindows = { ...state.openWindows };
+  const closedWindows = { ...state.closedWindows };
+  const tabId = tab.tabId;
+  const lasttabIndex = openWindows[targetWindow].tabs.findIndex(tab => {return tab === tabId});
```
Comment (joshuef) [inline app/reducers/windows.ts]: `lastTabIndex`

**Comment by joshuef** [comment]
@manav wasn't https://github.com/maidsafe/safe_browser/issues/599 to be sorted by this issue?

- [x] this does fix: https://github.com/maidsafe/safe_browser/issues/636

I would also say this issue can be the final nail in the coffin over here too: https://github.com/maidsafe/safe_browser/issues/557

**Comment by joshuef** [comment]
@manavbp select nextTab/previousTab (from menu, or via shortcut) is not working

**Comment by joshuef** [comment]
Also having many tabs open and switching between them causes issues where the active tab is no longer rendered

**Comment by joshuef** [comment]
I can sort the tab switch/rendering issue by switching active tabs to use `z-index` away from `visibility`. 

(Though I feel like that's going backwards..., as that was an olde issue)

**Comment by joshuef** [comment]
closing as this is superseded by #751

**Pull request closed without merging**